static/app/components/dropdownMenuV2.tsx

```diff
@@ -0,0 +1,301 @@
+   * See:
+   *   https://react-spectrum.adobe.com/react-aria/useOverlay.html
+   *   https://react-spectrum.adobe.com/react-aria/useOverlayPosition.html
+   */
+  const overlayRef = useRef(null);
+  const {overlayProps} = useOverlay(
+    {
+      onClose: closeRootMenu,
+      shouldCloseOnBlur,
+      isDismissable,
+      isOpen: true,
+    },
+    overlayRef
+  );
+  const {overlayProps: positionProps, placement: placementProp} = useOverlayPosition({
+    targetRef: triggerRef,
+    overlayRef,
+    offset,
+    crossOffset,
+    placement,
+    containerPadding,
+    isOpen: true,
+  });
+
+  /**
+   * Store whether this menu/submenu is the current focused one, which
+   * in a nested, tree-like menu system should be the leaf submenu.
+   * This information is used for controlling keyboard events.
+   * See: modifiedMenuProps below.
+   */
+  const [hasFocus, setHasFocus] = useState(true);
+  useEffect(() => {
+    /**
+     * A submenu is a leaf when it does not contain any expanded submenu.
+     * This logically follows from the tree-like structure and single-selection
+     * nature of menus.
+     */
+    const isLeafSubmenu = !stateCollection.some(node => {
+      const isSection = node.hasChildNodes && !node.value.isSubmenu;
+      /**
+       * A submenu with key [key] is expanded if
+       * state.selectionManager.isSelected([key]) = true
+       */
+      return isSection
+        ? [...node.childNodes].some(child =>
+            state.selectionManager.isSelected(`${child.key}`)
+          )
+        : state.selectionManager.isSelected(`${node.key}`);
+    });
+    setHasFocus(isLeafSubmenu);
+  }, [state.selectionManager.selectedKeys]);
+  /**
+   * Menu props from useMenu, modified to disable keyboard events
+   * if the current menu does not have focus.
+   */
+  const modifiedMenuProps = {
+    ...menuProps,
+    ...(!hasFocus && {
+      onKeyUp: () => null,
+      onKeyDown: () => null,
+    }),
+  };
+
+  /**
+   * Render a single menu item
+   */
+  const renderItem = (node: Node<MenuItemProps>, isLastNode: boolean) => {
+    return (
+      <MenuItem
+        node={node}
+        isLastNode={isLastNode}
+        state={state}
+        onClose={closeRootMenu}
+        closeOnSelect={closeOnSelect}
+      />
+    );
+  };
+
+  /**
+   * Render a submenu whose trigger button is a menu item
+   */
+  const renderItemWithSubmenu = (node: Node<MenuItemProps>, isLastNode: boolean) => {
+    const trigger = ({props: submenuTriggerProps, ref: submenuTriggerRef}) => (
+      <MenuItem
+        ref={submenuTriggerRef}
+        renderAs="div"
+        node={node}
+        isLastNode={isLastNode}
+        state={state}
+        isSubmenuTrigger
+        {...submenuTriggerProps}
+      />
+    );
+
+    return (
+      <MenuControl
+        items={node.value.children as MenuItemProps[]}
+        trigger={trigger}
+        menuTitle={node.value.submenuTitle}
+        placement="right top"
+        offset={-4}
+        crossOffset={-8}
```

Comment:
what are these numbers?

Reply:
this is terminology they created to account for different menu placements. offset is the spacing between the menu and its trigger along the main axis (when alignment = "top" or "bottom", the main axis is y/vertical, when it's "left" or "right", the main axis is x/horizontal). crossOffset is the offset along the cross axis, which is perpendicular to the main axis.

Reply:
so these numbers are for submenus only, which have `placement="right top"`, so the main axis is horizontal. `offset={-4}` means that submenus are moved to the left by 4px, and `crossOffset={-8}` means they're moved up by 8px.

<img width="295" alt="Screen Shot 2022-01-28 at 2 21 17 PM" src="https://user-images.githubusercontent.com/44172267/151629494-41824e1e-8131-4cc6-a300-23f4b9c9bd39.png">